### PR TITLE
[fix] Fix perf regression caused by MoE autotuner when using DeepEPLowLatency

### PR DESCRIPTION
**Pull request by jinyangyuan-nvidia**
## Summary by CodeRabbit

* **New Features**
  * Enhanced expert model operator to support additional tuning options when all-to-all communication is enabled.

* **Bug Fixes**
  * Improved input validation and handling for expert model operations, ensuring stricter checks and more robust behavior.

* **Refactor**
  * Adjusted internal parameter handling and input preparation for expert model tuning and execution.
  * Changed the order of returned token bucket sizes in utility functions to ascending order for consistency.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py [hunk 1]
```diff
@@ -141,24 +141,45 @@ def fused_moe(
...
     min_latency_mode: bool = False,
     tune_max_num_tokens: int = 8192,
+    total_valid_tokens: Optional[int] = None,
+    original_top_k: Optional[int] = None,
 ) -> List[torch.Tensor]:
 
     tuner = AutoTuner.get()
     MoERunner.refine_tuning_config(tune_max_num_tokens)
 
+    if enable_alltoall:
```
[the remark this answers is not in the document]
Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py:151]: When using DeepEP, `enable_alltoall` is expected to be `True` only in the generation phase. We usually set the environment variable `TRTLLM_DEEP_EP_TOKEN_LIMIT` to be the maximum number of tokens in the generation phase to ensure it.

Because `enable_alltoall` is `False` in the warmup, the inputs to the tuner are modified here when `enable_alltoall` is `True` so that the inputs are equivalent to the ones used in the warmup.

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py [hunk 2]
```diff
@@ -141,24 +141,45 @@ def fused_moe(
...
+        if input.shape[0] >= total_valid_tokens:
+            tuner_input = input[:total_valid_tokens].contiguous()
+        else:
+            tuner_input = torch.cat([
+                input,
+                torch.zeros(total_valid_tokens - input.shape[0],
+                            *input.shape[1:],
+                            dtype=input.dtype,
+                            device=input.device)
+            ])
```
Comment (yuantailing) [inline tensorrt_llm/_torch/custom_ops/torch_custom_ops.py:163]: Will you call the `contiguous` or `cat` at each iteration?

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py [hunk 3]
```diff
@@ -437,6 +437,17 @@ def forward_chunk(
...
+        # Prepare additional information for profiling in case padding is applied in all-to-all
+        if use_all_to_all:
+            if all_rank_num_tokens is not None:
+                total_valid_tokens = sum(all_rank_num_tokens)
+            else:
+                total_valid_tokens = x.shape[0] * self.mapping.tp_size
+            original_top_k = token_selected_slots.shape[1]
+        else:
+            total_valid_tokens = None
+            original_top_k = None
```
Comment (yuantailing) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py:450]: 1. The computation of `total_valid_tokens` seems wrong. It's much larger than the number of real received tokens per rank, especially when EP is large. The number `x.shape[0] * self.mapping.tp_size` is also padded. If a padded number is acceptable, what is the essential difference between this and DeepEPLowLatency paded num (`#local experts * EP size * all_rank_max_num_tokens`)? It's only a quantitative change, so I think this PR does not make sense.
2. Is `total_valid_tokens` only a size hint used for tuning? If so, please rename the variable to reflect that.

**Review by yuantailing — CHANGES_REQUESTED** [review]

File: tensorrt_llm/_torch/utils.py [hunk 4]
```diff
@@ -239,7 +239,7 @@ def get_last_power_of_2_num_tokens_buckets(max_num_tokens) -> List[int]:
     while m >= 1:
         num_token_buckets.append(m)
         m //= 2
-    return tuple(num_token_buckets)
+    return tuple(num_token_buckets[::-1])
```
[the remark this answers is not in the document]
Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/utils.py:242]: https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/autotuner.py#L532 assumes ascending ordering. The original ordering here is descending, which is incorrect.

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py (hunk 2 above)
Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: Good point. The code has been modified.

Although the contiguous layout is not required now, the `contiguous` is still added to avoid potential bugs in the future by ensuring that the inputs to the MoE profiler are identical for `use_all_to_all = True` and `use_all_to_all = False`. The `contiguous` has been moved to the `forward` function of `MoERunner`, which is only called when cache misses in the warmup phase.

The `cat` in the `else` branch is never called because `input.shape[0]` is always larger than `total_valid_tokens` when `use_all_to_all = True`. It is added previously for logical completeness. This branch has been removed in the updated commit.

File: tensorrt_llm/_torch/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuantailing) [reply tensorrt_llm/_torch/utils.py:242]: Should we also change `get_power_of_2_num_tokens_buckets` above?

[the remark this answers is not in the document]
Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/utils.py:242]: Thanks, done.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py (hunk 3 above)
Comment (yuantailing) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py]: I want to understand the mechanism of tuning.
For example, given `tuner_num_tokens` and `top_k`, is each expert MatMul with `tuner_num_tokens` tokens? Where is `top_k` used?
If so, should we use the "math expectation" for `tuner_num_tokens` which is `x.shape[0] * TP size * top_k / num_slots`?

Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py]: No. `tuner_num_tokens` is `x.shape[0]` (the number of tokens after all-gather) if not using all-to-all. `top_k` is used inside `fused_moe` to expanding inputs with `x.shape[0]` tokens to inputs with `x.shape[0] * top_k` tokens. The number of tokens for each expert is stochastic during profiling. The average number of tokens for each expert is `x.shape[0] * top_k / num_slots`.

Comment (yuantailing) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py]: Thanks for explanation! I saw `token_selected_experts` is generated in `prepareFakeRouterBuffers` for each token respectively, so setting `total_valid_tokens` to the same as AllGather is reasonable. I have approved the PR and I hope a brief reason can be written in the code comment.

**Review by yuantailing — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py (hunk 3 above)
Comment (yuantailing) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_wide_ep.py]: I think the value of `token_selected_experts` takes no effect because the number of tokens used for tuning is generated by `get_last_power_of_2_num_tokens_buckets`. Only `top_k` influences tuning results, right?

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py [hunk 5]
```diff
@@ -88,7 +88,7 @@ def forward(
     ):
         x, fc1_expert_weights, fc1_expert_biases, fc2_expert_weights, fc2_expert_biases = inputs
         self.fused_moe_runner.run_gemm_profile(
-            x,
+            x.contiguous(),
```
Comment (yuantailing) [inline tensorrt_llm/_torch/custom_ops/torch_custom_ops.py:91]: Will the time of `contiguous()` counted in the gemm1/gemm2 time? If so, I think it's better to keep the original logic because this PR does not change whether x is contiguous.

Comment (jinyangyuan-nvidia) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: Good point. Done.

**Review by hyukn — APPROVED** [review]
